packages/firestore/src/remote/online_state_tracker.ts

```diff
@@ -0,0 +1,158 @@
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { OnlineState } from '../core/types';
+import * as log from '../util/log';
+import { assert } from '../util/assert';
+
+const LOG_TAG = 'OnlineStateTracker';
+
+// To deal with transient failures, we allow multiple stream attempts before
+// giving up and transitioning to Offline.
+const MAX_WATCH_STREAM_FAILURES = 2;
+
+// To deal with stream attempts that don't succeed or fail in a timely manner,
+// we have a maximum timeout we'll wait for the stream to either succeed or fail
+// MAX_WATCH_STREAM_FAILURES times, else we revert to OnlineState.Offline.
+const MAX_WATCH_STREAM_TIMEOUT_MS = 10 * 1000;
+
+/**
+ * A component used by the RemoteStore to track the OnlineState (that is,
+ * whether or not the client as a whole should be considered to be online or
+ * offline), implementing the appropriate heuristics.
+ *
+ * In particular, when the client is trying to connect to the backend, we
+ * allow up to MAX_WATCH_STREAM_FAILURES within MAX_WATCH_STREAM_TIMEOUT_MS for
+ * a connection to succeed. If we have too many failures or the timeout elapses,
+ * then we set the OnlineState to Offline, and the client will behave as if
+ * it is offline (get()s will return cached data, etc.).
+ */
+export class OnlineStateTracker {
+  /** The current OnlineState. */
+  private state = OnlineState.Unknown;
+
+  /**
+   * A count of consecutive failures to open the stream. If it reaches the
+   * maximum defined by MAX_WATCH_STREAM_FAILURES, we'll set the OnlineState to
+   * Offline.
+   */
+  private watchStreamFailures = 0;
+
+  /**
+   * A timer that elapses after MAX_WATCH_STREAM_TIMEOUT_MS, at which point we
+   * revert to OnlineState.Offline without waiting for the stream to actually
+   * fail (MAX_WATCH_STREAM_FAILURES times).
+   */
+  // tslint:disable-next-line:no-any setTimeout() type differs on browser / node
+  private watchStreamTimer: any = null;
+
+  /**
+   * Whether the client should log a warning message if it fails to connect to
+   * the backend (initially true, cleared after a successful stream, or if we've
+   * logged the message already).
+   */
+  private shouldWarnClientIsOffline = true;
+
+  constructor(private onlineStateHandler: (onlineState: OnlineState) => void) {}
+
+  /**
+   * Called by RemoteStore when a watch stream is started.
+   *
+   * It sets the OnlineState to Unknown and starts a MAX_WATCH_STREAM_TIMEOUT_MS
+   * timer if necessary.
+   */
+  handleWatchStreamStart(): void {
+    this.setAndBroadcast(OnlineState.Unknown);
+
+    if (this.watchStreamTimer === null) {
+      this.watchStreamTimer = setTimeout(() => {
```

Comment:
Given all the timer gunk we have now I wonder if it would make sense to either add methods to the async queue to just do this for us or add a special clock/delay type that you could pass in to the constructor to allow us to test these kinds of things directly.